JSON test builder

jannikluhn opened the pull request:
Referencing #332 

This is an unfinished first iteration of a JSON state test builder with the goal to explore if this approach makes sense. For sharding some changes in the test runner and the JSON format are required, so to keep things simple I started with replicating main chain tests.

I posted a usage example the resulting JSON file [here](https://gist.github.com/jannikluhn/208707c73e96bf9f7ad8bdbff6154e88).

For comparison, from what I understood the original JSON tests are generated from a hand written JSON file (or automatically generated using some pattern) which is subsequently "filled" (i.e. expanded and the resulting state is calculated). Compared to my approach this is safer as it doesn't use the tested code to generate the tests (the example above for instance uses the `BaseTransaction` class). But it makes writing tests more cumbersome and less clear (especially when encoding transaction data for, e.g., user account transactions).

Anyways, I'm very interested in your opinions and other ideas on this, @pipermerriam @hwwhww @NIC619.

#### Cute Animal Picture

![put a cute animal picture link inside the parentheses](https://pixabay.com/get/ea34b50720f51c22d2524518a33219c8b66ae3d018b616449df5c57d/lop-eared-314881_1280.jpg)

pipermerriam commented on evm/utils/test_builder.py:
```diff
@@ -0,0 +1,309 @@
+        account_state_dict.setdefault("balance", 0)
+        account_state_dict.setdefault("nonce", 0)
+        account_state_dict.setdefault("code", b"")
+        account_state_dict.setdefault("storage", OrderedDict())
```
This will mutate the original state dictionary which is very likely to cause un-expected errors.  Can you change this to a non-mutating pattern?

pipermerriam commented on evm/utils/test_builder.py:
```diff
@@ -0,0 +1,309 @@
+    def __init__(
+        self,
+        name,
+        *,
```
I'm not familiar with this pattern.  What's it do?

jannikluhn replied:
It requires the following arguments to be given by keyword (https://www.python.org/dev/peps/pep-3102/). This has the side effect that you can add parameters with no default after other keyword arguments (`transaction` and `secret_key`).

pipermerriam replied:
I love this so much.  I'd like to backfill this pattern onto a bunch of places in py-evm and also the p2p code (cc @gsalgado).

pipermerriam commented on evm/utils/test_builder.py:
```diff
@@ -0,0 +1,309 @@
+        self.forks = forks
+
+        # pre state
+        self.env = merge(DEFAULT_ENV, env or OrderedDict())
```
You could just change `OrderedDict()` to `{}` here for slightly improved readability.  Same with other places that you use an empty ordered dict (unless your relying on the underlying `merge` to return an `OrderedDict` at which point please ignore (does it do that?).

jannikluhn replied:
Hmm, I assumed it does, but now I checked and I'm wrong... I tried to use `OrderedDict`s to keep a sensible order in the final JSON file (pre before transaction before post). But apparently the `ethereum/tests` standard is ordering alphabetically, so I don't think it's worth the effort. I'm gonna switch to plain dictionaries.

pipermerriam replied:
In this same vein, we *should* use `sort_keys=True` if this actually ends up dumping JSON strings at any point.

pipermerriam commented on evm/utils/test_builder.py:
```diff
@@ -0,0 +1,309 @@
+        return pre_state
+
+    def to_dict(self):
+        d = OrderedDict()
```
`eth-utils` has an `eth_utils.to_ordered_dict` which will let you write this using the `yield` pattern (if you're inclined to).

pipermerriam commented on evm/utils/test_builder.py:
```diff
@@ -0,0 +1,309 @@
+    return new_state
+
+
+def without_duplicates(l):
```
Available from [`toolz.unique`](http://toolz.readthedocs.io/en/latest/api.html)

pipermerriam commented on evm/utils/test_builder.py:
```diff
@@ -0,0 +1,309 @@
+    return list(OrderedDict.fromkeys(l))
+
+
+def flatten(l):
```
Available from [`toolz.concat`](http://toolz.readthedocs.io/en/latest/api.html#toolz.itertoolz.concat) though it doesn't return a `list` type but rather the return type from `itertools.chain.from_iterable`.

jannikluhn commented on evm/utils/test_builder.py:
```diff
@@ -0,0 +1,309 @@
+    return force_text(version)
+
+
+env_formatter = curried.apply_formatters_to_dict({
```
I love the formatters, they made the code much cleaner! The only one I missed was a dict key formatter (basically what `storage_formatter` and `state_formatter` do manually now), but I'm not sure if it's a good idea to add them to the library as it could destroy uniqueness of keys if one's not careful.

pipermerriam replied:
You're looking for `eth_utils.apply_key_map`

jannikluhn commented:
I started over completely:

1) Test generation is now split into two steps:
  - test specification in high level "filler" files
  - "compilation" of the filler file to the actual test file
2) removed classes, the JSON file itself (or rather, the corresponding dict) is generated step by step in a functional style with some helper functions

The motivation for 1) was to resemble the existing process of JSON test generation over at ethereum/tests as closely as possible. This should make integration with their and other tooling easier (e.g. filler files could be generated with py-EVM and then filled with different implementations to check that the state roots match).

I may have gone overboard with 2 a bit (it's just so much fun!), but in general it makes sense I think. The dict stores exactly the same data as the class, but with less clutter (got rid of these huge argument lists as well). Also, this should be more flexible and allow for manual modification if necessary.

As always, here's a usage example, with resulting filler and test file: https://gist.github.com/jannikluhn/5a8e6cd0e85e9f27c9679d2458fe8fe0

pipermerriam commented on evm/utils/test_builder/normalization.py:
```diff
@@ -0,0 +1,103 @@
+    """
+    d = {}
+    for state_item in state_list:
+        d = assoc_in(d, state_item[:-1], state_item[-1])
```
*functional nonsense*

```python
return cytoolz.merge(*(
    assoc_in({}, state_item[:-1], state_item[-1])
    for state_item
    in state_list
))
```

jannikluhn replied:
I like it, that's nicer (but needs `deep_merge` instead of `merge`). With 
```
deep_merge(*(
    assoc_in({}, state_item[:-1], state_item[-1]) if isinstance(state_item, Mapping) else state_item
    for state_item
    in state_list
))
```
this would be even more flexible (then one can switch from the list to dict definition at any level:
```
[
    ("0xaabb", "balance", 3),
    ("0xaabb", "storage", {
        4: 5,
    }),
    {"0xbbcc", {
        "balance": 6,
        "nonce": 7
    }
]
```

pipermerriam commented on evm/utils/test_builder/normalization.py:
```diff
@@ -0,0 +1,103 @@
+    """
+    d = {}
+    for state_item in state_list:
+        d = assoc_in(d, state_item[:-1], state_item[-1])
```
It *might* be good to pre-validate that `state_list` is well formed.  Otherwise this is going to produce some very confusing results that might be hard to track down.  Similarly, enforcing uniqueness of all `state_item[:-1]` would fit into that validation to ensure that no duplicates are present.

pipermerriam commented on evm/utils/test_builder/test_builder.py:
```diff
@@ -0,0 +1,218 @@
+@curry
+def pre_state(pre_state, filler):
+    assert len(filler) == 1
+    test_name = next(iter(filler.keys()))
```
These two lines are duplicated a few times.  Maybe write a small utility function  **or** change the signature to expect two arguments `pre_state(pre_state, test_name, state)`

pipermerriam commented on evm/utils/test_builder/builder_utils.py:
```diff
@@ -0,0 +1,74 @@
+from eth_keys import keys
+
+
+def merge_nested(*dicts):
```
*functional nonsense*

Slightly more concise and uses fully non-mutation based approach

```python
from cytoolz import merge_with

def merge_if_dicts(values):
    if all(isinstance(item, Mapping) for item in values):
        return merge_with(merge_if_dicts, *values)
    else:
        return values[-1]

def deep_merge(*dicts):
    return merge_with(merge_if_dicts, *dicts)
```

jannikluhn replied:
👍 for `deep_merge` over `merge_nested`, as its analogous to `copy.deepcopy`

pipermerriam commented:
Could we get a small `README.md` document somewhere in the fillers that documents how this stuff works at some basic level of granularity.  Doesn't have to be utterly thorough documentation but something to get others started building on this API.

carver commented on evm/utils/fixture_tests.py:
```diff
@@ -287,37 +290,34 @@ def generate_fixture_tests(metafunc,
+        is_main_transaction = True
     else:
-        raise KeyError("transaction missing secret key or vrs values")
+        is_main_transaction = False
```
I think it will help prevent future errors if we validate the shard transaction fields. Something like:

```py

def is_valid_shard_transaction(transaction):
    return all(key in transaction for key in {'shardID', 'accessList'})

if 'secretKey' in transaction:
...
elif is_valid_shard_transaction(transaction):
     is_main_transaction = False
else:
    raise ValueError("transaction must be a signed main chain transaction, or a shard transaction.")

jannikluhn commented:
Updates:

- rebased
- paygas tests fixed (they pass now when copied to the fixture directory)
- more strict normalization (missing and additional keys in dictionary should be catched now)
- https://github.com/ethereum/py-evm/pull/350/commits/65d7b24d718b1aa1238ba72a385d16baa93498e7: fixed shard VM instantiation which failed due to the differences between blocks and collations. I don't particularly like my change of `persist_block_to_db`, but it's the least invasive way to make it work (and I assume that the chain class has to be revised anyway to support the different types of clients).

Please review one (hopefully) last time.

Also, @NIC619 could you have a look at https://github.com/jannikluhn/py-evm/blob/aa7343e5d16ddf0292f0fd06d878f70651228133/tests/fillers/vm_fillers/paygas_fillers.py and check if we need additional test cases for PAYGAS?

pipermerriam commented on evm/db/chain.py:
```diff
@@ -403,10 +403,11 @@ def persist_block_to_db(self, block):
                 self._add_transaction_to_canonical_chain(transaction_hash, header, index)
 
         # Persist the uncles list
-        self.db.set(
```
Opened https://github.com/ethereum/py-evm/issues/410 to track this issue.

pipermerriam commented on evm/utils/test_builder/README.md:
```diff
@@ -0,0 +1,183 @@
+# JSON test builder
+
+
+The JSON test builder found in `evm.utils.test_builder.test_builder` is a library that
```
I recently introduced `evm.tools` into the module.  I think that `test_builder` should probably be moved there.

pipermerriam commented on evm/utils/test_builder/README.md:
```diff
@@ -0,0 +1,183 @@
+  `"expectedPeriodNumber"` | the period number in which the collation is supposed to be included
+  `"periodStartHash"`      | the hash of the block starting the current period.
+
+Use `setup_main_filler` or `setup_sharding_filler` with the same signature, but correct
```
This sentence appears to be truncated.

pipermerriam commented on evm/utils/test_builder/builder_utils.py:
```diff
@@ -0,0 +1,126 @@
+    merge_with,
+)
+from eth_utils import (
+    force_text,
```
`force_text` shouldn't be used.  It's been/being deprecated and removed.  If you need to coerce things from `bytes -> text` you should use `eth_utils.to_text` in combination with `eth_utils.text_if_str`

```python
from eth_utils.curried import text_if_str
from eth_utils import to_text

force_text = text_if_str(to_text)
```

The behavior isn't quite identical, but I think it's identical in all the ways that matter.  Note, that it **will not** be able to handle any arbitrary bytestring since not all 0-255 byte values are `utf8` encodable.

carver replied:
Or, if you know the input will always be a `bytes` value, then `to_text(bytes_val)` is sufficient.

pipermerriam commented on evm/utils/test_builder/builder_utils.py:
```diff
@@ -0,0 +1,126 @@
+
+
+def is_cleanly_mergable(*dicts):
+    if len(dicts) <= 1:
```
This function, while not a public API might benefit from a docstring since it's not entirely clear what *cleanly mergable* means without reading the implementation.  A human readable description of this function would go a long way to helping future readers understand it's intended functionality.

pipermerriam commented on evm/utils/test_builder/formatters.py:
```diff
@@ -0,0 +1,133 @@
+import eth_utils.curried
+
+
+int_to_hex = compose(encode_hex, int_to_big_endian)
```
this can be replaced with `eth_utils.to_hex`

pipermerriam commented on evm/utils/test_builder/normalization.py:
```diff
@@ -0,0 +1,278 @@
+        assert is_cleanly_mergable(*state_dicts)
+        state_dict = deep_merge(*state_dicts)
+    else:
+        assert False
```
It'd be nice to get a human readable error here that exposed the allowed types and the type that was observed.

pipermerriam commented on evm/utils/test_builder/normalization.py:
```diff
@@ -0,0 +1,278 @@
+            for state_item
+            in state_definition
+        ]
+        assert is_cleanly_mergable(*state_dicts)
```
It'd be nice to get an exception here that showed what key or maybe keys where found to not be cleanly mergable.

pipermerriam commented on evm/utils/test_builder/normalization.py:
```diff
@@ -0,0 +1,278 @@
+        assert False
+
+    second_layer_keys = concat(d.keys() for d in state_dict.values())
+    assert all(key in ["balance", "nonce", "storage", "code"] for key in second_layer_keys)
```
*minor performance thing*

- casting `second_layer_keys` to a `set` will remove duplicates so you don't check the same key multiple times.
- the set of allowed keys *should* be a `set` since we're using it purely to check for presence within the set.
- this could be done in one go with `bad_keys = set(seen_keys).difference({"balance", "nonce", ...})`

I'd recommend changing to use the `bad_keys = ` pattern because then you can raise an explicit error showing what *extra* keys were found.

pipermerriam commented on evm/utils/test_builder/test_builder.py:
```diff
@@ -0,0 +1,371 @@
+    "gasPrice": 0,
+    "nonce": 0,
+    "secretKey": decode_hex("0x45a915e4d060149eb4365960e6a7a45f334393093061116b197e3240065ff2d8"),
+    "to": to_canonical_address("0x0f572e5295c57f15886f9b263e2f6d2d6c7b5ec6"),
```
Did this address come from anywhere in particular?

jannikluhn replied:
It's used in a lot of existing JSON tests (https://github.com/ethereum/tests/search?utf8=%E2%9C%93&q=0x0f572e5295c57f15886f9b263e2f6d2d6c7b5ec6&type=), but I guess they just generated it randomly.

pipermerriam commented on evm/vm/forks/sharding/collations.py:
```diff
@@ -26,6 +26,8 @@ def __init__(self, header, transactions=None, witness_nodes=None):
         if witness_nodes is None:
             witness_nodes = []
 
+        self.transaction_fee_sum = 0
```
This doesn't appear to be used (as researched by a quick ctrl+f)

NIC619 replied:
`block.transaction_fee_sum`(or `collation.transaction_fee_sum`) is used to record the accumulated transaction fee paid by each transaction that's included in the block(collation).

```python
def add_transaction(self, transaction, computation, block):
    …
    # Calculate transaction fee
    transaction_fee, _ = computation.compute_transaction_fee_and_refund()
    # Bookkeep this transaction fee
    block.transaction_fee_sum += transaction_fee
    ...
```

pipermerriam replied:
Ah, I see.

I've opened #412 to track the task of trying to find a new way to do this.

I need to dig in and understand a bit better why this is there in the first place, but I suspect we can find a different mechanism.

NIC619 commented:
@jannikluhn What do you think about adding a negative gas price test case?

jannikluhn commented:
> @jannikluhn What do you think about adding a negative gas price test case?

The EVM doesn't know negative numbers, so we don't have to (and can't, even) write such a test here. But it's something the Solidity/Vyper guys should do in their tests.

jannikluhn commented on evm/rlp/headers.py:
```diff
@@ -273,3 +288,21 @@ def from_parent(cls,
+        return self.__class__(**{
+            field_name: getattr(self, field_name)
+            for field_name
+            in tuple(zip(*self.fields))[0]
```
Looks like `unzip` is only in `toolz`, not `cytoolz`, unfortunately. But I'll change the `tuple(..)[0]` to `cytoolz.first(..)` which is arguably a bit nicer.

pipermerriam replied:
Nice to know that `first` works on generator/iterator types.

jannikluhn commented:
Addressed the feedback and made some minor changes to make the tests pass. Will merge when tests are green (hope I caught everything this time).
